type_decoders/README.md

```diff
@@ -0,0 +1,787 @@
+```
+
+`isAnyOf()` accepts an array of decoders and attempts to decode a provided value using each of them, in order, returning the first successful result or `DecoderError` if all fail.
+
+### isChainOf()
+
+```ts
+interface IChainOfDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isChainOf<T extends [unknown, ...unknown[]], R = ChainOfDecoderReturn<T>, I = DecoderInputType<T[0]>>(
+  decoders: { [I in keyof T]: Decoder<T[I], DecoderInputType<T[SubtractOne<I>]>> },
+  options?: IChainOfDecoderOptions
+): Decoder<R, I>;
+
+function isChainOf<T extends [unknown, ...unknown[]], R = ChainOfDecoderReturn<T>, I = DecoderInputType<T[0]>>(
+  decoders: { [I in keyof T]: Decoder<T[I], DecoderInputType<T[SubtractOne<I>]>> | PromiseDecoder<T[I], DecoderInputType<T[SubtractOne<I>]>> },
+  options?: IChainOfDecoderOptions
+): PromiseDecoder<R, I>;
+```
+
+`isChainOf()` accepts an array of decoders and attempts to decode a provided value using all of them, in order. The successful output of one decoder is provided as input to the next decoder. `isChainOf()` returns the `DecoderSuccess` value of the last decoder in the chain or `DecoderError` on the first failure.
+
+### isObject()
+
+```ts
+interface IObjectDecoderOptions<T> {
+  msg?: DecoderErrorMsgArg;
+  keyMap?: { [P in keyof T]?: string | number };
+}
+
+function isObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> }, options?: IObjectDecoderOptions<T>): Decoder<T> ;
+function isObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> | PromiseDecoder<T[P]> }, options?: IObjectDecoderOptions<T>): PromiseDecoder<T> ;
+```
+
+`isObject()` accepts a `key: Decoder` argument object and returns a new decoder that will verify that an input is a non-null object, that the input has the keys specified in the argument object, and that the key values pass the relevant decoder provided by the argument object. On `DecoderSuccess`, the a new object is returned and the key-values of that object are provided by output from the argument decoders.
+
+### isExactObject()
+
+```ts
+interface IExactObjectDecoderOptions<T> {
+  msg?: DecoderErrorMsgArg;
+  keyMap?: { [P in keyof T]?: string | number };
+}
+
+function isExactObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> }, options?: IExactObjectDecoderOptions<T>): Decoder<T> ;
+function isExactObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> | PromiseDecoder<T[P]> }, options?: IExactObjectDecoderOptions<T>): PromiseDecoder<T> ;
+```
+
+`isExactObject()` is the same as `isObject()`, except the input object cannot have any excess properties.
+
+### isDictionary()
+
+```ts
+interface IDictionaryDecoderOptions<V> {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isDictionary<R, V = unknown>(decoder: Decoder<R, V>, options?: IDictionaryDecoderOptions<V>): Decoder<R[], V>;
+function isDictionary<R, V = unknown>(decoder: PromiseDecoder<R, V>, options?: IDictionaryDecoderOptions<V>): PromiseDecoder<R[], V>;
```

Comment:
It looks like the return types here have a typo — should they be `{ [key: string]: R }` instead of `R[]`?

Reply:
Good catch